### PR TITLE
[ConstraintSystem] Rework overload ranking based on Sendable conformances

### PR DESCRIPTION
**Pull request by xedin**
- Increase score if conformance constraint finds missing conformances
  
  Add a new score kind to denote presence of missing conformances -
  `SK_MissingSynthesizableConformance`.

  The types with missing conformances are allowed to appear in a
  valid solution but such solutions should be ranked lower comparing
  to solutions fewer or without them.

- Specify whether specialization check allows missing conformances or not

  We have two places where `CompareDeclSpecializationRequest` is used:

  - A performance optimization that compares two generic overloads;
  - Solution ranking that checks all of the selected overloads against
    another solution.

  The former can allow missing conformances and shouldn't prevent
  the solver from checking overloads that differ on `Sendable`
  (because there is no information about what is passed as arguments)
  but the latter, since it has a solution, should prefer Sendable
  overloads over non-Sendable ones if possible
  (i.e. `init<T: Sendable>(_: T)` is a subtype of `init<T>(_: T)`).

Resolves: rdar://117227549

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test macOS platform

File: include/swift/Sema/ConstraintSystem.h [hunk 1]
```diff
@@ -988,8 +988,12 @@ enum ScoreKind: unsigned int {
   /// ambiguity tie-breakers should go after this; anything else
   /// should be added above.
   SK_UnappliedFunction,
+  /// A type with a missing conformance(s) that has be synthesized
+  /// or diagnosed later, such types are allowed to appear in
+  /// a valid solution.
+  SK_MissingSynthesizableConformance,
```
Comment (hborla) [inline include/swift/Sema/ConstraintSystem.h:994]: Should this go before `SK_UnappliedFunction`?

**Review by hborla — APPROVED** [review]

File: include/swift/Sema/ConstraintSystem.h (hunk 1 above)
Comment (xedin) [reply include/swift/Sema/ConstraintSystem.h:994]: I think this is lower impact than unapplied function (property with a missing conformance is still better than unapplied function?) and for ranking we’ve got a flag and a check in compare solutions.

**Comment by xedin** [comment]
@swift-ci please test

File: include/swift/Sema/ConstraintSystem.h (hunk 1 above)
Comment (xedin) [reply include/swift/Sema/ConstraintSystem.h]: We discussed this offline and came to the conclusion that it's better to move it up and filter first on Sendability.

**Pull request merged**